pymodbus/client/mixin.py

```diff
@@ -500,3 +502,66 @@ def read_device_information(
         return self.execute(
             pdu_mei.ReadDeviceInformationRequest(read_code, object_id, **kwargs)
         )
+
+    # ------------------
+    # Converter methods
+    # ------------------
+
+    class DATATYPE(Enum):
+        """Datatype enum for convert_* calls."""
+
+        INT16 = ("h", 1)
+        UINT16 = ("H", 1)
+        INT32 = ("i", 2)
+        UINT32 = ("I", 2)
+        INT64 = ("q", 4)
+        UINT64 = ("Q", 4)
+        FLOAT32 = ("f", 2)
+        FLOAT64 = ("d", 4)
+        STRING = ("s", 0)
+
+    @classmethod
+    def convert_from_registers(
+        cls, registers: list[int], data_type: DATATYPE
+    ) -> Union[int, float, str]:
+        """Convert registers to int/float/str.
+
+        :param registers: list of registers received from e.g. read_holding_registers()
+        :param data_type: data type to convert to
+        :returns: int, float or str depending on "to_type"
+        :raises ModbusException: when size of registers is not 1, 2 or 4
+        """
+        byte_list = bytearray()
+        for x in registers:
+            byte_list.extend(int.to_bytes(x, 2, "big"))
+        if data_type == cls.DATATYPE.STRING:
+            if byte_list[-1:] == b"\00":
+                byte_list = byte_list[:-1]
+            return byte_list.decode("utf-8")
+        if len(registers) != data_type.value[1]:
+            raise ModbusException(
+                f"Illegal size ({len(registers)}) of register array, cannot convert!"
+            )
+        return struct.unpack(f">{data_type.value[0]}", byte_list)[0]
+
+    @classmethod
+    def convert_to_registers(
+        cls, value: Union[int, float, str], data_type: DATATYPE
+    ) -> list[int]:
+        """Convert int/float/str to registers (16/32/64 bit).
+
+        :param value: value to be converted:
+        :param data_type: data type to convert to
+        :returns: List of registers, can be used directly in e.g. write_registers()
+        """
+        if data_type == cls.DATATYPE.STRING:
+            byte_list = value.encode()  # type: ignore[union-attr]
```

Comment:
@alexrudd2 Here is a (in my mind) a good explanation why mypy is a real PITA. This is perfectly safe code, but mypy do not recognise that.

But admitted mypy have other advantages, so I live with the ignores 🥲

Reply:
In this case, I think `if isinstance(value, str):` satisfies mypy by narrowing the type with no ignore.

https://mypy.readthedocs.io/en/latest/type_narrowing.html#type-narrowing-expressions


The docstring for `param: data_type` isn't clear.  Would it be better as 'data type to be encoded as registers'?

Also, it makes no difference but `Union[float, int, str]` can be written as `Union[float, str]`

Reply:
We accept float, int and str that's why I added all 3, a way to tell the user what is allowed.

I will change the doc of param data_type later.

I will have to test if is instance works....I suspect that
```
  isinstance(cls.DATATYPE.STRING) ==  isinstance(cls.DATATYPE.IN16)
```
are true, and thus it will not work.

Reply:
> We accept float, int and str that's why I added all 3, a way to tell the user what is allowed.

Yes, the method accepts all 3 types ("broad").  But `encode()` should only work on `str` ("narrow").  I will prepare a PR to demonstrate.